### PR TITLE
Isolate side effects in graphics functions

### LINKED ISSUE #227: Blank `Rplots.pdf` generated when running snapshot tests interactively from RStudio
To reproduce the issue:

```r
usethis::create_package("zzz/")
usethis::use_testthat()
```

Run `usethis::use_r("myfun")` and put this function into it:

```
#' @export
myfun <- function(path) {
  head(datasets::iris) |>
    r2rtf::rtf_body() |>
    r2rtf::rtf_encode() |>
    r2rtf::write_rtf(file = path)
}
```

Run `usethis::use_test("myfun")` to create this test:

```r
test_that("myfun snapshot", {
  path <- tempfile(fileext = ".rtf")
  myfun(path)

  local_edition(3)
  expect_snapshot_file(path, "myfun.rtf")
})
```

Run `devtools::document()` and `devtools::load_all()`.

Clicking the "Run Tests" button or the "Test" button in the Build panel in RStudio will generate a blank, redundant `tests/testthat/Rplots.pdf`, while no such `Rplots.pdf` is generated when running `devtools::test()` or `devtools::test_active_file()` in the R console.

Reproduced under macOS, Windows, and Posit Cloud.

### PR DESCRIPTION
**Pull request by nanxstats**
Fixes #227 

This PR introduces `safe_par()` and `safe_strwidth()` as alternatives to their original graphics functions. These wrappers will be evaluated in a temporary BMP device, to avoid generating `Rplots.pdf` when running r2rtf code under certain context - such as clicking the test buttons in RStudio environments, even for third-party packages.

@yihui I'm not sure if this is the optimal solution so I'm open to suggestions.

Using these wrappers can be be 10x slower than calling the graphics functions directly:

```r
microbenchmark::microbenchmark(
  r2rtf:::safe_strwidth("Hello world!", units = "inches", cex = 1, font = 1, family = "sans"),
  graphics::strwidth("Hello world!", units = "inches", cex = 1, font = 1, family = "sans"),
  times = 5000
)

#>    min     lq      mean median     uq      max neval cld
#> 44.526 46.945 53.097288 47.806 49.692 4399.300  5000   a
#>  2.747  2.952  3.199353  3.198  3.362   16.482  5000   b
```

### CODE REVIEW

File: R/rtf_strwidth.R [hunk 1]
```diff
@@ -131,3 +132,37 @@ rtf_strwidth <- function(tbl) {
...
+safe_strwidth <- function(text, ...) with_bmp(graphics::strwidth(text, ...))
+
+#' Evaluate expression within a BMP device and close it on exit
+#'
+#' See why: <https://github.com/Merck/r2rtf/issues/227>
+#'
+#' @noRd
+with_bmp <- function(expr) {
+  expr <- substitute(expr)
+  grDevices::bmp(file_null())
```
Comment (elong0527) [inline R/rtf_strwidth.R:153]: Just for my understanding. Why would bmp() device used instead of pdf()

File: R/rtf_strwidth.R [hunk 2]
```diff
@@ -131,3 +132,37 @@ rtf_strwidth <- function(tbl) {
...
+        grDevices::dev.off(current)
+        if (prev != current) grDevices::dev.set(prev)
+      }
+    },
+    add = TRUE
+  )
+  eval(expr, envir = parent.frame())
+}
+
+file_null <- function() if (.Platform$OS.type == "windows") "nul:" else "/dev/null"
```
Comment (elong0527) [inline R/rtf_strwidth.R:168]: Could we use "tempfile" to avoid a OS dependent logic?

File: R/rtf_strwidth.R [hunk 3]
```diff
@@ -131,3 +132,37 @@ rtf_strwidth <- function(tbl) {
...
+      current <- grDevices::dev.cur()
+      if (current > 1) {
+        prev <- grDevices::dev.prev(current)
+        grDevices::dev.off(current)
+        if (prev != current) grDevices::dev.set(prev)
+      }
+    },
+    add = TRUE
+  )
+  eval(expr, envir = parent.frame())
```
Comment (yihui) [inline R/rtf_strwidth.R:165]: I don't think you need `substitute()` + `eval()`, but can just take advantage of R's lazy evaluation of arguments, i.e., simply put `expr` at the end, and R will evaluate it at that time.

```suggestion
  grDevices::bmp(file_null())
  on.exit(
    {
      current <- grDevices::dev.cur()
      if (current > 1) {
        prev <- grDevices::dev.prev(current)
        grDevices::dev.off(current)
        if (prev != current) grDevices::dev.set(prev)
      }
    },
    add = TRUE
  )
  expr
```

**Review by yihui — COMMENTED** [review]
I think a more elegant and faster solution is to use the null pdf device, i.e.,

```r
opts <- options(device = function(...) pdf(NULL, ...))
on.exit(options(opts), add = TRUE)
```

File: R/rtf_strwidth.R (hunk 2 above)
Comment (yihui) [reply R/rtf_strwidth.R:168]: I guess writing to the null device is often faster than a real file.

**Pull request closed without merging**